### DISCUSSION TITLE
Bookmarks / Outline support - quick fix

### DISCUSSION THREAD

**Discussion by m-holger** [qpdf/qpdf#1244] [Ideas]
As part of the `pages epic` much more comprehensive support for outlines is planned. However, to provide comprehensive support is quite complex and there are a number of preliminary steps that need to be implemented before tackling this task. On the other hand, the question of providing more limited support crops up regularly, e.g. most recently in https://github.com/qpdf/qpdf/issues/343#issuecomment-2229312854 and #1243.

This raises the question whether it would be worthwhile to provide a more limited solution that could be implemented more quickly. In particular, would it be helpful to provide a facility to extract very basic information about the document outline to a file, and to replace the outlines in a PDF file with the content of such a file. This could be as simple as providing for each outline item only the title, level in the outline hierarchy, and the page number that the item points to.

The questions are:

- how useful would this be?
- what information should be included to make this useful (keeping in mind that the more is required, the smaller the chance that this could be provided quickly)?

**Comment by kenwdcao** [comment]
It is very helpful to have this feature. In my use case, I often need to combine a lot of pdfs. Each of them is a table, listing or figure, which has title. After combining them I need to create bookmarks using the original titles. In this case, because I have the number of pages of each individual pdf, it’s easy to create a list of bookmark titles and destination number in the combined pdf. It would be really helpful if this list (in predefined syntax) could be processed by QPDF to generate bookmarks. Support of multiple Indention levels is also helpful. I think it’s more or less similar to what pdftk does.

**Comment by foolishgrunt** [comment]
If it wasn't already apparent (that being my comment linked to up there), I would most definitely use a feature like this. I think the only critical parameters it would need to support are the three you mentioned: title, level, and page number. This would also bring it to feature parity with pdftk.

I don't know if this is relevant or useful, but I stumbled upon a [Gist](https://gist.github.com/bsidhom/dae50ecc0062a7a1202469860c8eea89?permalink_comment_id=5123120) yesterday from someone who has apparently found a way to use qpdf's current functions to insert an outline. Unfortunately it's a bit over my head.

**Reply by jberkenbilt** [reply]
I didn't read it thoroughly, but it looks like the right idea. My impulse at this stage would be to take an approach like that with an external tool that uses qpdf JSON to do this. I'd be reluctant to build an interim/partial solution into qpdf proper because then we're stuck supporting it indefinitely. If we do, we'll need to really make sure we do it in a manner we can live with and extend if the initial version lacks some functionality. I'd also point out that the Gist uses Python to interface with qpdf's JSON. I love seeing examples of applications that make use of qpdf's JSON capabilities. That said, if it's in Python and there's no constraint about additional packages, you could also use pikepdf to manipulate the PDF structure. I realize that doesn't help with it being over your head, but I can confirm that it would take something along those lines to do this. I don't imagine it could be much simpler. This is because outlines are a document-level construct that needs to know about destinations which in turn contain PDF object references to pages. This isn't too bad when you're just working with the outline hierarchy manually. Where it gets really hard is if you start splitting and merging and duplicating and resizing and combining pages. See #1104 for a much more in-depth discussion.

**Reply by foolishgrunt** [reply]
> That said, if it's in Python and there's no constraint about additional packages, you could also use pikepdf to manipulate the PDF structure.

Thanks for the tip, the [documentation for pikepdf](https://pikepdf.readthedocs.io/en/latest/topics/outlines.html) makes it looks pretty simple. I think even I could handle that! :)

That said, I was hoping to avoid having to install and call any separate program. (My interest in the Gist was mostly to glean principles that I might be able to re-implement.) My end goal is a Powershell script that I can distribute to my coworkers and have them run with minimal setup. If the choice is between A) my current solution with qpdf and pdftk, and B) a new solution with qpdf, python, and pikepdf, I think my current solution is preferable.

**Reply by m-holger** [reply]
I too did not read it thoroughly. My first impression is that the program deals with the more straight-forward part - reinserting the outlines, which means that we have control over how the destinations are implemented. While extracting from the existing file we will have to deal with whatever we find, whether we have direct destinations, named destinations or actions.

I do think the facility to extract and reinsert destinations will be required even when qpdf provides full support for outlines. Once you get beyond simple concatenation of files, there are so many possible ways of combining outlines that qpdf could not possibly cater for them all without making the CLI impossibly complex. The only option I can see is to give users the option to make final adjustments themselves.

**Reply by jberkenbilt** [reply]
Yes, I agree. Something like this will be essential. qpdf will never be able to figure out exactly what to do except in the most straightforward cases.

**Comment by m-holger** [comment]
Reflecting on this, I think the best format to provide support for the this would be a JSON  array of top-level outline items, where each item has an array of its children. This seems easy enough to manipulate both using manual editing and using scripts, and as far as I can see at the moment, the format can be extended in a back-ward compatible way to cope with future comprehensive outline support.

The alternative of representing  the hierarchy by an explicit `level` item would be less flexible in the long-term.

Are there any problems with this approach that I am missing?

**Example**

```json
[
  {
    "page": 1,
    "title": "A sample document",
    "kids": [
      {
        "page": 3,
        "title": "Chapter 1",
        "kids": [
          {
            "page": 5,
            "title": "Subchapter 1.1"
          }
        ]
      },
      {
        "page": 42,
        "title": "Chapter 2"
      }
    ]
  },
  {
    "page": 73,
    "title": "An appended document"
  }
]
```

**Syntax**

![Outlines](https://github.com/user-attachments/assets/688e31ae-ced2-4834-a974-e656b1fbd551)

**Reply by foolishgrunt** [reply]
Would this be agnostic with respect to the indent level? I think it would be, if I'm reading it correctly.

I would prefer this, as I think it would make scripted bookmark generation simpler.

**Reply by m-holger** [reply]
If you mean by agnostic that  `kids` arrays are unchanged if their level changes, either because you add an outer level or remove an intermediate level, than yes, the format is agnostic in that respect.

**Reply by foolishgrunt** [reply]
Sorry, what I mean is will
```
[
  {
    "page": 1,
    "title": "A sample document",
    "kids": [
      {
        "page": 3,
        "title": "Chapter 1",
        "kids": [
          {
            "page": 5,
            "title": "Subchapter 1.1"
...
```
be interpreted the same as
```
[
{
"page": 1,
"title": "A sample document",
"kids": [
{
"page": 3,
"title": "Chapter 1",
"kids": [
{
"page": 5,
"title": "Subchapter 1.1"
...
```

**Reply by m-holger** [reply]
Yes. In JSON, all white-space is the same. It does not matter whether you have a single space, a new-line, or a newline and 4 spaces. The indentations are purely there to make it more readable for humans. If you want to know what exactly counts as white-space and where it is required, have a look at json.org .

**Reply by foolishgrunt** [reply]
Then it's good for me!

**Reply by jberkenbilt** [reply]
> Are there any problems with this approach that I am missing?

Just keep in mind that bookmarks can have various types of destinations, not just page numbers (e.g. a region of a page, a page and a zoom level, etc.) so as long as we have a way to represent everything, something like this should be good. I'd suggest making it easy to represent easy things but having a way to capture whatever's in the PDF so that dumping the outlines into this format and reloading them would be lossless. If we allow the breadth of what the PDF spec allows, it would be possible for people to use this as an interim solution for n-up kinds of configurations as well, should that get added to qpdf via the pages epic before full support for outlines does.

**Reply by foolishgrunt** [reply]
@m-holger
For what it's worth, your proposed format looks very similar to what's currently in use by [pdfcpu](https://github.com/pdfcpu/pdfcpu). Here's a sample output .json file I get from running `& pdfcpu.exe bookmarks export input.pdf bookmarks.json`:
```
{
	"header": {
		"source": "A Sample Document",
		"version": "pdfcpu v0.11.0 dev",
		"creation": "2025-07-29 15:32:57 PDT",
		"creator": "Writer",
		"producer": "pdfcpu v0.11.0 dev"
	},
	"bookmarks": [
		{
			"title": "Section 1",
			"page": 1

		},
		{
			"title": "Section 2",
			"page": 2,
			"kids": [
				{
					"title": "Section 2.1",
					"page": 2,
					"kids": [
						{
							"title": "Section 2.1.1",
							"page": 2,
							"kids": [
								{
									"title": "Section 2.1.1.1",
									"page": 2
								},
								{
									"title": "Section 2.1.1.2",
									"page": 2
								},
							]
						},
						{
							"title": "Section 2.1.2",
							"page": 3,
							"kids": [
								{
									"title": "Section 2.1.2.1",
									"page": 3
								},
								{
									"title": "Section 2.1.2.2",
									"page": 3
								},
								}
							]
						}
					]
				}
			]
		},
		{
			"title": "Section 3",
			"page": 4
		}
	]
}
```
A notable shortcoming of this format (like @jberkenbilt alluded to) is that there doesn't seem to be proper handling for bookmarks pointing to intermediate lines on the page. For example, although the original PDF had the bookmarks for `Section 2.1.1.1` and Section `2.1.1.2` pointing to different lines on page 2, this is not captured in the .json output, so writing this exact same .json file back to the file with `& pdfcpu bookmarks import -r input.pdf bookmarks.json output.pdf` makes it so that all the bookmarks for page 2 now point to the beginning of that page.

But that's pretty manageable, for my use case. Rudimentary support is better than no support, yes?🤞

**Comment by SeeLook** [comment]
Is it possible to achieve bookmarks/outlines change from libqpdf (C++),
or tinkering with json is the only way as long as `page epic`  is parked for now?  

Looking how to handle outlines with qpdf brought me here.

**Comment by jberkenbilt** [comment]
It's always possible to use the C++ API, which gives you the ability to make arbitrary changes to the file. The issue is that the outline structure is a bit complex with bidirectional links and destinations stored at the document level but that point to the page, and so on. So you can manipulate outlines using the C++ API, but the library won't help you very much. The qpdf library does have some support for working with outlines, and there are even some examples in the examples directory that show you how to do it. It's just not a complete solution that does everything most people want.

**Reply by SeeLook** [reply]
Thank You for the response.  
I know this bookmarks example but it is only about how to read them.  
This part I already did, I can grab bookmarks from will-to-be-merged PDF files  
into model structure and use it for UI.  

Now there is a dilemma how to store them all to output  file.  
I will dig qpdf code more, but any clue appreciated.  

BTW, this is the project: https://invent.kde.org/graphics/karp

**Reply by SeeLook** [reply]
Seems like this JSON solution from [gist](https://gist.github.com/bsidhom/dae50ecc0062a7a1202469860c8eea89?permalink_comment_id=5123120) is the only right direction for now  
but for my case - it has to be ported to C++.  
Because indeed, the library doesn't help with that.  
I'm not able to figure out how to do this with `QPDFObjectHandle` list of pages and other objects (outline dictionary).

**Reply by SeeLook** [reply]
Not that bad. I figured out how to do this without JSON:
```C++
// qpdf is already assembled output pdf before qpdf.write()
auto qpdfRoot = qpdf.getRoot();
auto outKey = qpdfRoot.getKey("/Outlines"s);
// store into std::pair list of every page object id (first) and generation (second)
std::vector<std::pair<int, int>> pageRefs;
for (auto &page : qpdf.getAllPages()) {
    pageRefs.push_back(std::pair<int, int>(page.getObjectID(), page.getGeneration()));
}
auto outlines = QPDFObjectHandle::newDictionary();
// with assert that /Outlines key already exists
qpdf.replaceObject(outKey.getObjectID(), outKey.getGeneration(), outlines);
int prevId = -1, prevGen = -1;

// iterate through nested bookmarks
// so let n is current bookmark node
// which has title, location point (x, y) and zoom
{
    auto bStream = qpdf.newStream();
    auto bDict = QPDFObjectHandle::newDictionary();
    bDict.replaceKey("/Count"s, QPDFObjectHandle::newInteger(0));
    bDict.replaceKey("/Title"s, QPDFObjectHandle::newUnicodeString(n->title().toStdString()));
    bDict.replaceKey("/Parent"s, outlines.getObj());
    if (prevGen > -1 && prevId > -1) {
        bDict.replaceKey("/Prev", qpdf.getObject(prevId, prevGen));
        auto prevObj = qpdf.getObject(prevId, prevGen);
        prevObj.replaceKey("/Next", qpdf.getObject(bStream.getObjectID(), bStream.getGeneration()));
    }
    // array for /Dest key
    auto destArr = QPDFObjectHandle::newArray();
    auto &p = pageRefs[n->pageNumber() - 1];
    destArr.appendItem(qpdf.getObject(p.first, p.second));
    destArr.appendItem(QPDFObjectHandle::newName("/XYZ"s));
    destArr.appendItem(QPDFObjectHandle::newReal(n->location().x()));
    destArr.appendItem(QPDFObjectHandle::newReal(n->location().y()));
    destArr.appendItem(QPDFObjectHandle::newReal(n->zoom()));
    bDict.replaceKey("/Dest"s, destArr);
    prevId = bStream.getObjectID();
    prevGen = bStream.getObjGen().getGen();
    qpdf.replaceObject(bStream.getObjectID(), bStream.getObjGen().getGen(), bDict);
    if (n == m_nodes.first())
        outlines.replaceKey("/First", qpdf.getObject(bStream.getObjectID(), bStream.getGeneration()));
    else if (n == m_nodes.last())
        outlines.replaceKey("/Last", qpdf.getObject(bStream.getObjectID(), bStream.getGeneration()));
}
outlines.replaceKey("/Count"s, QPDFObjectHandle::newInteger(bokmarksCount()));
```
This works for me.
